megatron/core/datasets/s3_indexed_dataset.py

```diff
@@ -0,0 +1,359 @@
+        We take the block with index (`offset` // `cache_size`) to refresh the
+        cache. If this new block still does not cover the requested span, we extend
+        it just enough to include `offset` + `size`.
+        """
+        if (
+            self._cache is not None
+            and offset >= self._cache_bytes_start
+            and offset + size <= self._cache_bytes_end
+        ):
+            return self._extract_from_cache(offset, size)
+
+        bytes_start = (offset // self._cache_nbytes) * self._cache_nbytes
+        assert bytes_start >= 0
+        assert offset >= bytes_start
+        bytes_end = max(bytes_start + self._cache_nbytes, offset + size)
+        assert bytes_end >= 1
+        self._cache = self._client.get_object(
+            Bucket=self._bucket,
+            Key=self._key,
+            # Subtract 1, because the end of Range is inclusive.
+            Range=f'bytes={bytes_start}-{bytes_end-1}',
+        )['Body'].read()
+        self._cache_bytes_start = bytes_start
+        self._cache_bytes_end = bytes_end
+        return self._extract_from_cache(offset, size)
+
+    def close(self):
+        self._client.close()
+
+
+def _download_file(s3_client, s3_path, local_path):
+    dirname = os.path.dirname(local_path)
+    os.makedirs(dirname, exist_ok=True)
+    parsed_s3_path = _parse_s3_path(s3_path)
+    s3_client.download_file(parsed_s3_path.bucket, parsed_s3_path.key, local_path)
+
+
+def _maybe_download_file(s3_path, local_path):
+    if torch.distributed.is_initialized():
+        rank = torch.distributed.get_rank()
+        # https://github.com/NVIDIA/Megatron-LM/blob/\
+        # 89574689447d694bb19dd86fc8a6153b4467ba9d/megatron/initialize.py#L232
+        local_rank = rank % torch.cuda.device_count()
+        world_size = torch.distributed.get_world_size()
+    else:
+        rank = 0
+        local_rank = 0
+        world_size = 1
+
+    s3_client = boto3.client("s3")
+
+    if (not os.path.exists(local_path)) and (rank == 0):
+        _download_file(s3_client, s3_path, local_path)
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    # If the `local_path` is in a file system that is not
+    # shared across all the ranks, then we assume it's in the
+    # host file system and each host needs to download the file.
+    if (not os.path.exists(local_path)) and (local_rank == 0):
+        _download_file(s3_client, s3_path, local_path)
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    # If the `local_path` still does not exist, then we assume
+    # each rank is saving to a separate location.
+    if not os.path.exists(local_path):
+        _download_file(s3_client, s3_path, local_path)
+
+    if torch.distributed.is_initialized():
+        torch.distributed.barrier()
+
+    assert os.path.exists(local_path)
+
+
+class S3IndexedDataset(torch.utils.data.Dataset):
```

Comment:
Can we make this class inherit from MMapIndexedDataset, keeping the length, get item, and properties, but overloading the s3 specific functions?

setstate and getstate could even be inherited if state were presumed defined in init

Reply:
After some consideration and after seeing this [refactor](https://github.com/NVIDIA/Megatron-LM/commit/23f3f55b894f6a290590cc49c04b53d9540a7e59), it seemed like the best way to integrate S3 data loading into IndexedDataset was to introduce an abstract _BinReader class and concrete _MMapBinReader, _FileBinReader and _S3BinReader subclasses, but let me know your thoughts on the design. I'm open to feedback.